### PR TITLE
Change types to base what `visitor` gets on `tree`

### PR DESCRIPTION
**Pull request by wooorm**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

While the signature `visitParents(tree, 'emphasis', (node: Emphasis) => void)` already works (that is, assuming that the tree could potentially include an emphasis node, so not hast, and that the tree does not include an `type: emphasis` which is otherwise different from `Emphasis`), the other signatures were not as smart.

Given

```
const root: Root = {...}
visitParent(root, (node: ?) => {})
```

`?` is now filled with any node that is `root` or (recursively) in `root.children`.

EDIT: this now also figures out the signature *with* a test!

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -48,9 +48,9 @@ export const EXIT = false
 export const visitParents =
   /**
    * @type {(
-   *   (<T extends Node>(tree: Node, test: T['type']|Partial<T>|import('unist-util-is').TestFunctionPredicate<T>|Array.<T['type']|Partial<T>|import('unist-util-is').TestFunctionPredicate<T>>, visitor: Visitor<T>, reverse?: boolean) => void) &
-   *   ((tree: Node, test: Test, visitor: Visitor<Node>, reverse?: boolean) => void) &
-   *   ((tree: Node, visitor: Visitor<Node>, reverse?: boolean) => void)
+   *   (<Needle extends Node>(tree: Node, test: Needle['type']|Partial<Needle>|import('unist-util-is').TestFunctionPredicate<Needle>|Array.<Needle['type']|Partial<Needle>|import('unist-util-is').TestFunctionPredicate<Needle>>, visitor: Visitor<Needle>, reverse?: boolean) => void) &
```
Comment (wooorm) [inline index.js:51]: `T -> Needle` is a better name.

I did try to use `Visitor<Extract<NodeInTree<Tree>, Needle>>` here but it did not work.

File: complex-types.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import type {Node, Parent} from 'unist'
+
+export type NodeInTree<T extends Node, A = void> = T extends Parent
+  ? T | NodeInTree<Exclude<T['children'][number], A | T>, A | T>
+  : T
```
Comment (wooorm) [inline complex-types.d.ts:5]: Open to different names.
Previously I used `SelfAndChildren`.

We could also extract this type. Either in a repo: `syntax-tree/typefest`? Or in `@types/unist`.

File: index.test-d.ts [hunk 3]
```diff
@@ -24,6 +19,41 @@ interface Element extends Parent {
...
+interface Emphasis extends Parent {
+  type: 'emphasis'
+  children: Phrasing[]
+}
+
+interface Text extends Literal {
+  type: 'text'
+  value: string
+}
+
```
Comment (remcohaszing) [inline index.test-d.ts:56]: I’m thinking it may be better to just use an existing AST for testing to ensure real world compatibility. Actually using `@types/mdast` here is probably even less effort.

File: index.test-d.ts [hunk 4]
```diff
@@ -109,3 +139,33 @@ visitParents(sampleTree, 'heading', (_) => [SKIP, 1])
...
+        {type: 'text', value: 'c'}
+      ]
+    }
+  ]
+}
+
+visitParents(typedTree, (_: Root | Flow | Phrasing) => {})
+const blockquote = typedTree.children[0]
+assert(blockquote.type === 'blockquote')
+visitParents(blockquote, (_: Flow | Phrasing) => {})
```
Comment (remcohaszing) [inline index.test-d.ts:168]: Wouldn’t it be better to use something like this?

```ts
visitParents(blockquote, (child) => {
  expectType<Flow | Phrasing>(child)
})
```

The same goes for 3 lines below.

File: complex-types.d.ts (hunk 2 above)
Comment (remcohaszing) [reply complex-types.d.ts]: What about `Descendant`? Naming is hard.

File: index.js (hunk 1 above)
Comment (remcohaszing) [reply index.js:51]: Maybe something like this? (I didn’t try it)

```ts
<Tree extends Node, Needle extends NodeInTree<Tree>>(tree: Tree, test: Needle['type']|Partial<Needle>|import('unist-util-is').TestFunctionPredicate<Needle>|Array.<Needle['type']|Partial<Needle>|import('unist-util-is').TestFunctionPredicate<Needle>>, visitor: Visitor<Needle>, reverse?: boolean) => void) &
```

File: index.test-d.ts (hunk 3 above)
Comment (wooorm) [reply index.test-d.ts:56]: Yeah maybe. I thought it was wise to keep this small and contained

File: index.test-d.ts (hunk 4 above)
Comment (wooorm) [reply index.test-d.ts:168]: Both work the same I believe. Your proposal is longer though. Do you have strong feelings about which to prefer?

File: complex-types.d.ts (hunk 2 above)
Comment (wooorm) [reply complex-types.d.ts]: https://github.com/syntax-tree/unist#descendant

`InclusiveDescendant` to be sure?

Comment (wooorm) [reply complex-types.d.ts]: We could add type utilities for several things in Unist to types/unist perhaps? 🤔

Comment (remcohaszing) [reply complex-types.d.ts]: Maybe. Although really `NodeInTree` isn’t that bad either actually.

File: index.test-d.ts (hunk 4 above)
Comment (remcohaszing) [reply index.test-d.ts:168]: I believe `expectType` is an actual type assertion, right? Whereas the following is also allowed when using the current syntax:

```ts
visitParents(blockquote, (_: Flow | Phrasing | string | true | 42) => {})
```

Comment (wooorm) [reply index.test-d.ts:168]: I guess your proposal protects against `any`s, so that’s nice

File: index.js (hunk 1 above)
Comment (wooorm) [reply index.js:51]: That does work fine, but the catch-all type (2nd overload) matches instead when the result would be `never`, thus yielding `Node` instead of `never` 🤔

Comment (wooorm) [reply index.js:51]: ohhh it works!!!

File: index.test-d.ts (hunk 4 above)
Comment (wooorm) [reply index.test-d.ts]: done!

File: test.js [hunk 5]
```diff
@@ -13,9 +17,15 @@ import remark from 'remark'
 import gfm from 'remark-gfm'
 import {visitParents, EXIT, SKIP, CONTINUE} from './index.js'
 
-const tree = remark().parse('Some _emphasis_, **importance**, and `code`.')
+/** @type {Root} */
 // @ts-expect-error: hush.
```
Comment (ChristianMurphy) [inline test.js:21]:
```suggestion
// @ts-expect-error: return type is known to be Root with children
```

File: index.js (hunk 1 above)
Comment (ChristianMurphy) [reply index.js:51]: Currently this compiles to
```ts
export const visitParents: (<Needle extends import('unist').Node>(
  tree: Node,
  test: any,
  visitor: Visitor<Needle>,
  reverse?: boolean | undefined
) => void) &
  (<Tree extends import('unist').Node>(
    tree: Tree,
    test: any,
    visitor: Visitor<any>,
    reverse?: boolean | undefined
  ) => void) &
  (<Tree_1 extends import('unist').Node>(
    tree: Tree_1,
    visitor: Visitor<any>,
    reverse?: boolean | undefined
  ) => void)
```

anything which gets `test` to a stricter type linked with `visitor` would be good.
Trying this with @remcohaszing's change gives me:

```ts
export const visitParents: (<
  Tree extends import('unist').Node<import('unist').Data>,
  Needle extends import('./complex-types').NodeInTree<Tree, void>
>(
  tree: Tree,
  test:
    | Needle['type']
    | Partial<Needle>
    | import('unist-util-is').TestFunctionPredicate<Needle>
    | (
        | Needle['type']
        | Partial<Needle>
        | import('unist-util-is').TestFunctionPredicate<Needle>
      )[],
  visitor: Visitor<Needle>,
  reverse?: boolean | undefined
) => void) &
  (<Tree_1 extends import('unist').Node<import('unist').Data>>(
    tree: Tree_1,
    test: Test,
    visitor: Visitor<import('./complex-types').NodeInTree<Tree_1, void>>,
    reverse?: boolean | undefined
  ) => void) &
  (<Tree_2 extends import('unist').Node<import('unist').Data>>(
    tree: Tree_2,
    visitor: Visitor<import('./complex-types').NodeInTree<Tree_2, void>>,
    reverse?: boolean | undefined
  ) => void)
```

Which is closer to what I expect :+1: 
The second overload still seems mildly disconcerting, `test` and `visitor` could still be disconnected in that case.
What case/edge-case is that overload aiming to cover?

Comment (ChristianMurphy) [reply index.js:51]: It's also not immediately clear to me what `Needle` refers to.
Maybe alliteration to needle in a haystack?

Maybe `NestedChildNode` or `NestedNode`?
Or an alternative playful name `MatryoshkaNode`

File: index.js [hunk 6]
```diff
@@ -48,9 +48,9 @@ export const EXIT = false
 export const visitParents =
```
Comment (ChristianMurphy) [inline index.js:48]: It seems that TSDoc description only comes through when attached to the const

```suggestion
/**
 * Visit children of tree which pass a test
 */
export const visitParents =
```

Comment (wooorm) [reply index.js:56]: fixed!

**Comment by wooorm** [comment]
Fix the types. The “catch-all” overload is no longer needed.
It was needed because `Needle` (the node type to be found) could only be explicitly given (so through TypeScript), making the “catch-all” overload required for JavaScript.
I’ve now changed the behavior of the types to:
- find all node types in the tree
- extract node types that match the given test in the type system.

Take for example the docs in [How to narrow `Node`s in TypeScript](https://unifiedjs.com/learn/recipe/narrow-node-typescript/), on lists:

```ts
if (is<List>(node, 'list')) {
  // If we’re here, node is List.
  //
  // `'list'` is compared to `node.type` to make sure they match.
  // `true` means a match, `false` means no match.
  //
  // `<List>` tells TypeScript to ensure `'list'` matches `List.type` and that
  // if `'list'` matches both `node.type` and `List.type`, we know that node is
  // `List` within this if condition.
}
```

(although this code is for `Is<>`, not for `Visit<>`).
The `<List>` is no longer needed explicitly in `Visit`, it is inferred.

**Comment by ChristianMurphy** [comment]
Nice!
Is `visit`ing a generic `Node` still supported? Or going forward will be always need to start with a more specific tree/node type?

**Comment by wooorm** [comment]
If you mean that `visit<SomeNodeType>(anObjectThatLooksLikeANode, 'some-node-type', (node: SomeNodeType) => {})` still works, then yes.
Because that “replaces” `Tree` from `anObjectThatLooksLikeANode` with `SomeNodeType` instead, and assuming that `'some-node-type'` matches it, will pass that node type as `node`.

There are some complexities with for example `visit({type: 'heading', depth: 1, children: []}, 'heading')` erroring, because `depth` and `children` do not exist on `Node`.
Or `visit(typedTree, {type: 'heading'}, (node: Heading) => {})` erroring, because TS `{type: 'heading'}` is turned into `{type: string}` which then yields all nodes instead.

File: index.js [hunk 7]
```diff
@@ -45,22 +45,27 @@ export const SKIP = 'skip'
  */
 export const EXIT = false
 
+/**
+ * Visit children of tree which pass a test
+ *
+ * @param tree Abstract syntax tree to walk
+ * @param test Test node, optional
+ * @param visitor Function to run for each node
+ * @param reverse Fisit the tree in reverse, defaults to false
```
Comment (ChristianMurphy) [inline index.js:54]:
```suggestion
 * @param [reverse=false] Visit the tree in reverse order
```

File: index.test-d.ts [hunk 8]
```diff
@@ -36,76 +88,99 @@ expectError(visitParents())
...
+
+  const child = paragraph.children[1]
+
+  if (is<Emphasis>(child, 'emphasis')) {
+    visitParents(child, 'blockquote', (node) => {
+      // `blockquote` does not exist in phrasing.
+      expectType<never>(node)
+    })
+  }
+}
```
Comment (ChristianMurphy) [inline index.test-d.ts:186]: it may be good to have one or two type tests showing it works with plain `Node`/`Parent` https://github.com/syntax-tree/unist-util-visit-parents/pull/10#issuecomment-888366338

**Review by ChristianMurphy — APPROVED** [review]

File: index.js (hunk 7 above)
Comment (wooorm) [reply index.js:54]: the `[]` brackets at this point don’t affect my editor (Atom). Do you see them having effect?

Comment (ChristianMurphy) [reply index.js:54]: :thinking: strange I don't see them in VSCode either, maybe because there is no `{boolean}`?
It could also be fine to keep the default in the description

```suggestion
 * @param reverse Visit the tree in reverse order, defaults to false
```

Comment (wooorm) [reply index.js:54]: I think the reason is because there are overloads, so values and defaults of things don’t matter (they differ per overload).

- The top-most one describes all overloads, so that’s why I don’t describe types and explicitly in prose write whether things are optional
- One level deeper is TS only. Docs are impossible.
- Level 3 is for the actual code and is used to check JS, so descriptions and prose don’t matter

File: index.test-d.ts (hunk 8 above)
Comment (wooorm) [reply index.test-d.ts:208]: Added some tests for implicit trees.

Unfortunately I was wrong. It doesn’t work :'(

```
visit<SomeNodeType>(anObjectThatLooksLikeANode, 'some-node-type', (node: SomeNodeType) => {})
```

Does not work, because `anObjectThatLooksLikeANode` is probably not assignable to `SomeNodeType`

**Comment by wooorm** [comment]
@ChristianMurphy Still good to go? I think this should be a major?

**Review by ChristianMurphy — APPROVED** [review]
Looks good! :+1: 
And yes, with some potentially breaking cases around generic `Node`, semver major sounds more appropriate.
